File: spx-gui/src/utils/xgo-executor/frameworks/tutorial.ts
import type { XGoExecutor, XGoFramework } from '../core/executor'

export type TutorialProgress = { completed: number; total: number }

export type TutorialCapabilities = {
  showMessage: (content: string) => void | Promise<void>
  readCode: () => string | Promise<string>
  setProgress: (progress: TutorialProgress) => void | Promise<void>
}

export function createTutorialFramework(capabilities: TutorialCapabilities): XGoFramework {
  return {
    name: 'tutorial',
    capabilities: {
      showMessage: (request) => capabilities.showMessage((request as { content: string }).content),
      readCode: async () => ({ code: await capabilities.readCode() }),
      setProgress: (request) => capabilities.setProgress(request as TutorialProgress)
    }
  }
}

export function dispatchTutorialSubmit(executor: XGoExecutor, submission: string): Promise<void> {
  return executor.dispatchEvent('submit', { submission })
}

File: tools/xgoexec/frameworks/tutorial/framework/tutorial.go
package tutorial

import "github.com/goplus/builder/tools/xgoexec/core"

const GopPackage = true

type Course struct {
	onStart func()
}

type CourseProto interface {
	MainEntry()
	Start()
}

type messageRequest struct {
	Content string `json:"content"`
}

type codeResult struct {
	Code string `json:"code"`
}

type submissionResult struct {
	Submission string `json:"submission"`
}

type progressRequest struct {
	Completed int `json:"completed"`
	Total     int `json:"total"`
}

func (p *Course) OnStart(handler func()) { p.onStart = handler }

func (p *Course) Start() {
	if p.onStart != nil {
		p.onStart()
	}
}

func (p *Course) ShowMessage(content string) {
	mustCallCapability("showMessage", messageRequest{Content: content}, nil)
}

func (p *Course) ReadCode() string {
	var result codeResult
	mustCallCapability("readCode", nil, &result)
	return result.Code
}

func (p *Course) WaitForSubmit() string {
	var result submissionResult
	mustCallCapability("waitForSubmit", nil, &result)
	return result.Submission
}

func (p *Course) SetProgress(completed, total int) {
	mustCallCapability("setProgress", progressRequest{Completed: completed, Total: total}, nil)
}

func Gopt_Course_Main(course CourseProto) {
	course.MainEntry()
	course.Start()
}

func mustCallCapability(name string, request, result any) {
	if err := core.CallCapability(name, request, result); err != nil {
		panic(err)
	}
}
